File: Tests/iOS Tests/PNUnsubscribeTests.m
//
//  PNUnsubscribeTests.m
//  PubNub Tests
//
//  Created by Jordan Zucker on 6/16/15.
//
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import <PubNub/PubNub.h>

#import "PNBasicSubscribeTestCase.h"

@interface PNUnsubscribeTests : PNBasicSubscribeTestCase <PNObjectEventListener>
@end

@implementation PNUnsubscribeTests

- (BOOL)isRecording{
    return NO;
}

- (void)setUp {
    [super setUp];
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertNotNil(status);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.category, PNConnectedCategory);
        NSArray *expectedPresenceSubscriptions = @[@"a", @"a-pnpres"];
        XCTAssertEqualObjects(status.subscribedChannels, expectedPresenceSubscriptions);
        XCTAssertEqual(status.operation, PNSubscribeOperation);
        NSLog(@"timeToken: %@", status.currentTimetoken);
        XCTAssertEqualObjects(status.currentTimetoken, @14355436534569417);
        XCTAssertEqualObjects(status.currentTimetoken, status.data.timetoken);
        
    };
    self.didReceiveMessageAssertions = ^void (PubNub *client, PNMessageResult *message) {
        PNStrongify(self);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertEqualObjects(client.uuid, message.uuid);
        XCTAssertNotNil(message.uuid);
        XCTAssertNil(message.authKey);
        XCTAssertEqual(message.statusCode, 200);
        XCTAssertTrue(message.TLSEnabled);
        XCTAssertEqual(message.operation, PNSubscribeOperation);
        NSLog(@"message:");
        NSLog(@"%@", message.data.message);
        XCTAssertEqualObjects(message.data.message, @"*************** 5657 - 2015-06-28 19:07:34");
        [self.subscribeExpectation fulfill];
    };
    [self PNTest_subscribeToChannels:@[@"a"] withPresence:YES];
    
}

- (void)testUnsubscribeWithPresence {
    PNWeakify(self);
    self.didReceiveStatusAssertions = ^void (PubNub *client, PNSubscribeStatus *status) {
        PNStrongify(self);
        XCTAssertNotNil(client);
        XCTAssertNotNil(status);
        XCTAssertEqualObjects(self.client, client);
        XCTAssertEqual(status.category, PNDisconnectedCategory);
        XCTAssertFalse(status.isError);
        XCTAssertEqual(status.statusCode, 200);
        XCTAssertEqual(status.operation, PNUnsubscribeOperation);
        [self.unsubscribeExpectation fulfill];
    };
    [self PNTest_unsubscribeFromChannels:@[@"a"] withPresence:YES];
}

@end

File: Tests/iOS Tests-Swift/PNUnsubscribeTests.swift
//
//  PNUnsubscribeTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 10/20/15.
//
//

import UIKit
import XCTest

import Foundation

class PNUnsubscribeTests: PNBasicSubscribeTestCase {
    
    override func isRecording() -> Bool {
        return false
    }
    
    let subscriptionChannels = ["a"]
    
    
    override func setUp() {
        super.setUp()
        
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNStatus) -> (Void) in
            
            XCTAssertEqual(self.client, client)
            XCTAssertNotNil(status)
            XCTAssertFalse(status.error)
//            XCTAssertTrue(status.category == .PNConnectedCategory)
            
            
            let expectedChannels: [String] = ["a", "a-pnpres"]
            
            let subscribedStatus = status as? PNSubscribeStatus
            
            if (subscribedStatus != nil) {
                var resultSet: Set<String> = []
                for element in (subscribedStatus?.subscribedChannels as? [String])! {
                    resultSet.insert(element)
                }
                
                let expectedChannelsSet: Set<String> = Set(expectedChannels)
                
                XCTAssertTrue(resultSet == expectedChannelsSet, "Subscribed channel groups list are not equal")
                
                XCTAssertEqual(subscribedStatus?.currentTimetoken, subscribedStatus?.data.timetoken)
                
                
                XCTAssertEqual(subscribedStatus?.currentTimetoken, NSDecimalNumber(string : "14356475647691168"))
                XCTAssertEqual(subscribedStatus?.currentTimetoken, subscribedStatus?.data.timetoken)
            }
            
            XCTAssertTrue(status.operation == .SubscribeOperation)
            
            self.subscribeExpectation.fulfill()
        }
        
        // TODO: investimate why it is not called
        
        self.didReceiveMessageAssertions = {(client: PubNub!, message: PNMessageResult!) -> (Void) in
            
            XCTAssertEqual(self.client, client)
            XCTAssertEqual(client.uuid(), message.uuid)
            XCTAssertNotNil(message.uuid)
            XCTAssertNil(message.authKey)
            XCTAssertEqual(message.statusCode, 200)
            XCTAssertTrue(message.TLSEnabled)
            XCTAssertTrue(message.operation == .SubscribeOperation)
            XCTAssertEqual(message.data.message as? String, "****........... 7161 - 2015-06-29 23:59:25")
        };
        
        self.PNTest_subscribeToChannels(self.subscriptionChannels, presense: true)
    }
    
    func testUnsubscribeWithPresence() {
        self.didReceiveStatusAssertions = {(client: PubNub, status: PNStatus) -> (Void) in
            
            XCTAssertNotNil(client)
            XCTAssertNotNil(status)
            XCTAssertEqual(self.client, client)
//            XCTAssertTrue(status.category == .PNDisconnectedCategory)
            XCTAssertFalse(status.error)
        //    XCTAssertEqual(status.statusCode, 200)
//            XCTAssertTrue(status.operation == .UnsubscribeOperation)
            self.unsubscribeExpectation.fulfill()
        }
        
        self.PNTest_unsubscribeFromChannels(self.subscriptionChannels, presense: true)
    }
}
